### ISSUE TITLE
508-defect-2 [SCREENREADER]: Alert icons MUST be implemented in a meaningful way for assistive technology

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#19674]
# [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

**Feedback framework**

- **❗️ Must** for if the feedback must be applied 
- **⚠️Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Description

The icons in the alerts must be implemented in a way that can be used or ignored by assistive technology. In this case, these icons mean "alert" to a sighted user, so they should convey that for non-sighted users, as well. This change applies to the alerts on:
- Intent to file
- Supporting evidence 

[This particular case is a little more challenging as the icons are css generated content, which means adding alt text will not have support on all browsers.](https://adrianroselli.com/2020/10/alternative-text-for-css-generated-content.html)

## Point of Contact

**VFS Point of Contact:** Josh

## Acceptance Criteria

* [ ] The icon itself is hidden from screen reader users (e.g. it should not announce "hollow square bullet" on JAWS)
* [ ] A text alternative is provided to screen reader users

## Environment

* Operating System: all
* Browser: all
* Screenreading device: all
* Server destination: production

### The usual approach
Typically, icon fonts on VA.gov are used via `<i>`. This is relatively more straightforward to fix as `aria-hidden` can be added to each `<i>` and a `<span class="sr-only">` can serve as the alternative text. See an example below pulled from another ticket:

#### Current code
```diff
<div class="usa-alert usa-alert-error background-color-only notice-marg-pad">
!  <i class="fa fa-exclamation-circle vads-u-margin-top--1 icon-base"></i>
  <div class="usa-alert-body">
    Facility Closed
  </div>
</div>
```

For each `<i>`, add attributes and screen reader only text:
- `aria-hidden="true"`
- `role="img"`
- <span class="sr-only">Alert: </span>
- <span class="sr-only"> temporarily due to covid-19</span>

#### Recommended enhancement
```diff
<div class="usa-alert usa-alert-error background-color-only notice-marg-pad">
+  <i  aria-hidden="true" role="img" class="fa fa-exclamation-circle vads-u-margin-top--1 icon-base"></i>
+  <span class="sr-only">Alert: </span>
  <div class="usa-alert-body">
    Facility Closed
    <span class="sr-only"> temporarily due to covid-19</span>
  </div>
</div>
```

### The tricky nature of css generated content
Unfortunately in this case, css generated content is being used to insert the icon. 

[W3C documents a way to add alt text...](https://www.w3.org/TR/css-content-3/#alt)

```html 
.new::before {
 content: url(./img/star.png) / "New!";
  /* or a localized attribute from the DOM: attr("data-alt") */
}
```

...but upon further research [it's spotty, and is not supported on Firefox and Safari](https://adrianroselli.com/2020/10/alternative-text-for-css-generated-content.html).

### Final recommendation
I'll need to do some more research, but at the moment I am currently leaning towards recommending **using an image + sr-only text similar to the usual approach documented above**. Several other experts in the field recommend this approach including Adrian Roselli and Stefan Judis.

## WCAG or Vendor Guidance (optional)

* [W3C WCAG 2.0 - Semantically identifying a font icon with role="img"](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA24.html)

* [W3C Icon Font with an On-Screen Text Alternative](https://www.w3.org/WAI/GL/wiki/Icon_Font_with_an_On-Screen_Text_Alternative) - Preferably, according to W3C guidance, the alert label would appear on-screen: "The objective of this technique is to show how to provide a visible, text alternative for an icon font that conveys information. Icon fonts are fonts that use the Private Use Area (PUA) of Unicode. Typically they are inserted in HTML via the CSS @font-face declaration and generated content property. Since they are vectors they are scalable and resolution-independent. Icon fonts can have 2 problems:   
    - Some people with disabilities may not use assistive technology (AT) and rely on on-screen text alternatives.
    - For those who do use AT, voicing of icon fonts may be inaccurate, nonsensical, redundant, or unpredictable.
    - To solve these 2 problems aria-hidden="true" is used so AT will ignore the icon. Then an on-screen text alternative is added to convey meaning to everyone."

## Screenshots or Trace Logs
We could in theory add in sr-only text within the heading, but JAWS will still announce the icon as "hollow square bullet" which is undesirable.
<img width="1508" alt="Screen Shot 2021-02-09 at 1 06 43 PM" src="https://user-images.githubusercontent.com/14154792/107410635-8fed2e80-6adb-11eb-8f10-e32980e799eb.png">

----

Screenshot of intent to file 
<img width="1424" alt="Screen Shot 2021-02-09 at 2 20 22 PM" src="https://user-images.githubusercontent.com/14154792/107416111-1dcc1800-6ae2-11eb-9252-d7fae9ed1131.png">
<img width="1411" alt="Screen Shot 2021-02-09 at 2 22 41 PM" src="https://user-images.githubusercontent.com/14154792/107416278-4bb15c80-6ae2-11eb-90da-6f1e0378ab61.png">

**Comment by caw310** [comment]
@Mottie , we're going to work on this in the next sprint.

**Comment by caw310** [comment]
Will address this issue when we build Alert Box web component. @bkjohnson will link tix that relate to this.

**Comment by bkjohnson** [comment]
Here is the in-progress PR that should address this: https://github.com/department-of-veterans-affairs/component-library/pull/74

**Comment by bkjohnson** [comment]
This is "solved" with the new `<va-alert>` web component which will be available in `vets-website` shortly. I put "solved" in quotes because VFS teams will still have to switch over to this new web component in order to get the enhancement.
